src/v/cluster/rm_stm.cc

```diff
@@ -1886,10 +1886,36 @@ void rm_stm::apply_data(model::batch_identity bid, model::offset last_offset) {
     }
 }
 
-void rm_stm::apply_checkpoint([
-  [maybe_unused]] const model::record_batch& batch) {}
+void rm_stm::apply_checkpoint(
+  [[maybe_unused]] const model::record_batch& batch) {}
 
-ss::future<> rm_stm::checkpoint_in_memory_state() const { co_return; }
+ss::future<model::record_batch> rm_stm::make_checkpoint_batch() {
+    iobuf key;
+    co_await serde::write_async(key, _insync_term);
+    iobuf val;
+    co_await serde::write_async(val, std::move(_mem_state));
```

Comment:
```
ss::future<model::record_batch> rm_stm::make_checkpoint_batch() {
...
    co_await serde::write_async(val, std::move(_mem_state));

```
i think that it needs to be formalized in some way that taking a checkpoint may or may not destroy _mem_state content. that doesn't seem like a casual side effect of building a checkpoint.